lib/cretonne/src/loop_analysis.rs

```diff
@@ -129,77 +169,162 @@ impl LoopAnalysis {
-                        // - None and node_loop is an new inner loop of lp
-                        // - Some(...) and the initial node_loop was a known inner loop of lp
-                        match node_loop_parent_option.expand() {
+                        // Now current_loop_parent is either:
+                        // - None and current_loop is an new inner loop of lp
+                        // - Some(...) and the initial node loop_id was a known inner loop of lp,
+                        //   in this case we have already visited it and we stop the DFS
+                        match current_loop_parent_option.expand() {
                             Some(_) => continue_dfs = None,
                             None => {
-                                if node_loop != lp {
-                                    self.loops[node_loop].parent = lp.into();
-                                    continue_dfs = Some(self.loops[node_loop].header)
-                                } else {
-                                    // If lp is a one-block loop then we make sure we stop
+                                let inner_loop = current_loop;
+                                let inner_loop_header = self.loops[inner_loop].header;
+                                if inner_loop == lp {
+                                    // If the inner loop is lp it means that we have reached a
+                                    // a block that has been previously tagged as part of lp so
+                                    // it has been visited by the DFS so we stop
                                     continue_dfs = None
+                                } else {
+                                    // Else we proceed to mark the inner loop as child of lp
+                                    // And continue the DFS with the inner loop's header
+                                    self.loops[inner_loop].parent = lp.into();
+                                    continue_dfs = Some(inner_loop_header);
+                                    // Here we perform a modification that's not related to
+                                    // the loop discovery algorithm but rather to the way we
+                                    // store the loop information. Indeed, for each Ebb we record
+                                    // the loop its part of and the last inst in this loop
+                                    // But here the Ebb is part of at least two loops, the child
+                                    // loop and the parent loop . So two cases arise:
+                                    // - either the entire ebb is part of the parent loop and this
+                                    // is fine;
+                                    // - either only part of the ebb is part of the parent loop and
+                                    // in that case we can't store the information of where does
+                                    // the parent loop stops.
```

Comment:
The word "either" should only appear once. One way to fix this would be to replace the second "either" with "or".

Also, remove the word "does" here.